### ISSUE TITLE
Prevent access to external files not in Config Files

### ISSUE THREAD

**Issue by muhammadaliog3** [JacquesCarette/Drasil#2193]
Configuration files are files that must be in same directory as executable in order for the executable to run. An example would be GlassBR that uses "SDF.txt" and "TSD.txt" in DataDefs.hs. While adding readme's to drasil in #2123 , config files were not stored anywhere, so they were added as a field in system information. They were added as a list of strings that the user must manually enter when the system information is defined. That leads to the following inconsistency:

Chunks can access files even if they are not mentioned in the config files field.

There are a few options
a) a warning is given if a file is accessed outside of the config files
b) an error   is given if a file is accessed outside of the config files
c) The config files field is automatically updated whenever the user accesses an external file.

**Comment by JacquesCarette** [comment]
Can you expand on this?  I mean, how would you detect such file access? You are right to point out this issue, I just want a better understanding before I give an opinion on a design.

**Comment by muhammadaliog3** [comment]
All calls to config files within GlassBr have similar structure, like the following which calls on the configfile "SDF.txt"
```haskell
tolPreEq = apply (asVC' interpY) [Str "SDF.txt", sy aspectRatio, sy sdfTol]
```
```interpY``` is the most interesting because it calls a function 
```haskell
readTable :: Func
readTable = funcData "read_table"
  "Reads glass ASTM data from a file with the given file name"
  [ singleLine (repeated [quantvar zVector]) ',',
    multiLine (repeated (map quantvar [xMatrix, yMatrix])) ','
  ]
```
Where singleLine and multiLine are from a module DataDesc.hs in drasil-code, according to the README in that module it "Defines types and contains helpers for dealing with data". Perhaps we could add some functionality to readTable or dataDesc that checks if the file being requested is in the configfiles in System Information.

**Comment by smiths** [comment]
@muhammadaliog3, I'm not sure I understand this issue.  Do we currently have a problem with Drasil code accessing config files that it shouldn't?

**Comment by muhammadaliog3** [comment]
No, we do not have a problem accessing config files currently. We can access config files just fine. This issue was created because we need to somehow record the names of the config files (ideally in system information).

**Comment by JacquesCarette** [comment]
And I think @muhammadaliog3 was pointing out something else: that we're currently quite insecure. It is potentially too easy in Drasil to create code that will access just about any file. We need to be more security-conscious in out codes.  Also, we shouldn't be burying the names of external resources so deeply inside code.

The above code to access "SDF.txt" is really bad in both those senses. We need to do better. Drasil is supposed to separate various concepts out more cleanly - and here we're seriously failing.

**Comment by smiths** [comment]
Thank you @muhammadaliog3 and @JacquesCarette.  I get it now.  Part of the problem is that the SDF.txt file is part of the design, not part of the requirements.  We do a good job of capturing, and generating documentation, for the requirements, but for the design decisions, we do not have a bridge.  The design decisions still had to be made to generate code, so at some point someone hard-coded them into Drasil.  That really isn't the same as separating the knowledge/decisions.

This is really part of a bigger issue that has come up in many guises, we need a systematic way to handle the design decisions.  As we refine from abstract requirements to concrete code, we need to make decisions.  Some of them we handle well, like the work that @bmaclach has done on his design language, but those are decisions embody a commonality analysis of all of our case study examples.  They do not cover example specific design knowledge, like how to do interpolation for GlassBR.

My thinking has been for a while that we want to do a "requirements" document for the design.  We talked about this idea briefly during Brooks's defence.  We could write a requirements document for an ODE solver and link those decisions to the needs of the physical problem that requires and ODE.  We would then be documenting what it means to be an ODE solver and the knowledge of what is an ODE.

Anyway, the current issue should just be about supplying the file information to GlassBR (and the other similar cases).  :smile:

**Comment by balacij** [comment]
It would be nice if #2149 could effectively remove the need for this ticket altogether by removing the 'external' part of the 'external sample input files'.
